### PR TITLE
Added StoreKeyConverter and StoreKeyConverterFactory interfaces (#910)

### PR DESCRIPTION
**Pull request by dharju**
StoreKeyConverter and StoreKeyConverterFactory interfaces should be added to facilitate real-time conversion of StoreKeys from one version to another.

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverter.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+   * Returns a mapping for every {@link StoreKey} in {@code input}. If a {@link StoreKey} is already in the format
+   * desired, returns the same {@link StoreKey} as the mapping. If no mapping exists, returns {@code null} as the
+   * mapping.
+   * (Note: TBD if we should return {@code null} if already in format requested but it is useful to diffrentiate
+   * b/w the cases)
+   * @param input the {@link StoreKey}s that need to be converted.
+   * @return a mapping for each {@link StoreKey} in the new format.
+   * @throws Exception that may be thrown when performing the conversion operation
+   */
+  public Map<StoreKey, StoreKey> convert(Collection<StoreKey> input) throws Exception;
```
Comment (pnarayanan) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverter.java]: Can the exception be more specific?

File: ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverter.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverter.java]: new line needed

File: ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverterFactory.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public interface StoreKeyConverterFactory {
+
+  /**
+   * Returns an instance of the {@link StoreKeyConverter} that the factory generates.
+   * @return an instance of {@link StoreKeyConverter} generated by this factory.
+   * @throws InstantiationException if the {@link StoreKeyConverter} instance cannot be created.
+   */
+  StoreKeyConverter getStoreKeyConverter() throws InstantiationException;
+}
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverterFactory.java]: new line

File: ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverterFactory.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * Usually called with the canonical class name and as such might have to support appropriate (multiple) constructors.
+ */
+public interface StoreKeyConverterFactory {
+
+  /**
+   * Returns an instance of the {@link StoreKeyConverter} that the factory generates.
+   * @return an instance of {@link StoreKeyConverter} generated by this factory.
+   * @throws InstantiationException if the {@link StoreKeyConverter} instance cannot be created.
+   */
+  StoreKeyConverter getStoreKeyConverter() throws InstantiationException;
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/store/StoreKeyConverterFactory.java]: `InstantiationException` is usually really hard to handle (should remove it from the old ones too). We should remove this and expect a `RuntimeException` if things go wrong in the factory.

**Pull request merged**